queries/just/injections.scm

```diff
@@ -0,0 +1,36 @@
+; The right side of =~ literals
+(regex_literal
+  (_) @injection.content
+  (#set! injection.language "regex"))
+
+; Default to bash highlighting for non-shebang recipes
+(recipe
+  (recipe_body) @injection.content
+  (#set! injection.include-children)
+  (#set! injection.language "bash"))
```

Comment:
Won't they *all* be injected with bash this way?

Reply:
I tried long and hard to find a way to match `(recipe_body)` only when it _doesn't_ have a `(shebang)` child node, but I couldn't. I found the `has-parent` and `has-ancestor` predicates in Neovim, but nothing similar for matching on child nodes.

When I tested this, it seemed to work anyway. I assumed the more specific queries below got precedence over this one, kind of like CSS. It didn't occur to me both could apply to the exact same region of text at the same time :thinking:

Do you know of a solution to this? Did I overlook anything in the regular query syntax, or are there any Neovim predicates that can solve this?

Reply:
There is something kind of weird with the grammars, TS and helix correctly capture the first line of the recipe as (shebang (lang)), but for some reason nvim doesn’t seem to capture that and just calls it part of the recipe body (meaning that e.g. `#!/use/bin/env python3` gets highlighted as a comment in the default language bash).

I just updated the queries yesterday to correctly handle defaults including `set shell` so that might be worth trying, but I haven’t yet gotten the chance to figure out this specific bug.

Reply:
>  TS and helix correctly capture the first line of the recipe as (shebang (lang))

Remember that TS (used to) and Helix (still) has a "first match wins" strategy, while Neovim (and TS and Helix will) has an "apply all and let the last match win" strategy. You cannot just copy queries.

Reply:
@clason In that case it sounds like this injection query is already acceptable for Neovim (?)

It applies a sane default first, then applies better matches if the shebang node is available. It seems to work perfectly.

Reply:
Not just conflicting, but random. Random highlight, random fold, random text objects, random indents, and so forth

Reply:
Yes. Last-wins only applies for highlight queries. For all other queries all are applied.

Reply:
No, _all_ matching queries (including highlight queries) are always applied, in order of their appearance in the query file. In the highlight case, this means that later matches "paint over" earlier ones and only "win" in this sense -- which is very weak, since if the first capture applies a bold font and the second capture applies a red color, you get bold red text.

Reply:
But here I'm talking about the effects of overlapping injections will have on other functionalities such as highlights, indents, and more though. Svelte and Vue were heavily affected by overlapped injections before, having Typescript code to be parsed as TS and JS, breaking everything.

So yes, please try your best to ensure no two injection queries have even a same matched node

Reply:
Sure, I was not saying anything about injections; just that there's no magic special-casing going on: all matching captures are always applied in all cases (and not just in Neovim).

Reply:
> @Hubro a couple days ago I added the named field needed to add `!shebang` to the default field which was needed for Helix, that may help here too https://github.com/IndianBoy42/tree-sitter-just/blob/5fe40d3622042c66465c4673b209a71e8376f902/queries/just/injections.scm#L17-L21. You will need to bump to the latest grammar sha.

That helps! :smile: